### DISCUSSION TITLE
Platformio not setting board define flag

### DISCUSSION THREAD

**Discussion by Deadeye5589** [earlephilhower/arduino-pico#1385] [Q&A] [answered]
Hello,

I'm trying to switch from the Arduino IDE to VSC and have been doing so for the past few hours. Crawaling from failure to failure. 
Right now I a have a project using a RPi Pico in combination with the GFX library from moononournation to drive an SPI display. This project compiles with the Arduino IDE. In VSC I get an error that a class definition for the display is missing. After some investigation I see that VSC skips the matching code block due to an compiler directive "if defined(TARGET_RP2040)". The definition of TARGET_RP2040 is nowhere to be found in the GFX libarary, hench my guess is, that this flag is normally set by the pico core. 

I have tried both variants of the core declarations from the platformio wiki article:

`[env:pico]
platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = pico
framework = arduino
board_build.core = earlephilhower`

and

`platform = https://github.com/maxgerhardt/platform-raspberrypi.git
board = rpipico
framework = arduino`

always with the same result of not being abel to compile. 

Now I'm asking you, if you have an idea why this flag is not set in VSC and how I can resvole this issue?

Best regards
Chris

**Comment by Deadeye5589** [comment]
Ok it seems that I have advanced a step
By adding "TARGET_RP2040" to the define section of the c_cpp_properties.json file I can get rid of the squiggles for the GFX library class definiton "Arduino_RPiPicoSPI".

Still unable to compile. Now running into an error: expected type-specifier before 'Arduino_RPiPicoSPI'
Which matches this discussion https://github.com/earlephilhower/arduino-pico/discussions/824
But I don't know how he could solve it sind the latest pico sdk version is 1.5.0 not 2.5.1
And for me it is not compiling on the regular Pi Pico and not just for a PicoW version.

**Reply by maxgerhardt** [reply]
`TARGET_RP2040`? That's sounds like like something from mbed-os as used in ArduinoCore-mbed. 

>This project compiles with the Arduino IDE. 

Which version of what core do you have installed there exactly?

**Reply by Deadeye5589** [reply]
Hmm interresting. The MBED Core is not supposed to be compatible with the GFX library from moononournation. So it would not make sense that there would be a compiler flag for it at all. 

My installation of the core is the V3.1.1. So the latest available. 
If I'm not mistaken this is where the compiler flag is normally set when using this core
platform.txt - Line 46 
-DTARGET_RP2040

**Reply by maxgerhardt** [reply, marked as answer]
Then that is indeed a bug since that's missing in 

https://github.com/earlephilhower/arduino-pico/blob/4f643d27d94cf8e0f4c507e79c15cd5a6f82c0e9/tools/platformio-build.py#L122-L129

But does it compile with with when you use `build_flags = -DTARGET_RP2040` in the `platformio.ini`? **All changes to the c_cpp_properties.json** will be ignored. Only the `platformio.ini` matters.

**Reply by Deadeye5589** [reply]
Indeed, there is some progess. There is a typo in your DTRAGET, but after correction the squiggles are gone and the project compiles. Unfortunately it wont't link. 
Now it need to figure out how this error is generated
collect2.exe: error: ld returned 1 exit status
*** [.pio\build\pico\firmware.elf] Error 1

**Reply by maxgerhardt** [reply]
The important part of the eror message is right before "ld returned 1 exit status". Something like, undefined reference, multiple refernence, ..

**Reply by Deadeye5589** [reply]
Ok, that is just marked as a warning, not an error.
Wow after almost 8 hours I seems that I can compile a project. As a novice this is really hard to learn an understand. 
Thank you for your support Max.

**Reply by maxgerhardt** [reply]
The linker error turned into a.. warning after doing nothing? Well okay I'm not gonna question this magic then.

A PR is open in https://github.com/earlephilhower/arduino-pico/pull/1386 to fix missing `TARGET_RP2040`.

**Reply by Deadeye5589** [reply]
Jepp, just a warning, nothing more. But now it seems to work. So "don't run a changing system" ;)

**Comment by Deadeye5589** [comment]
Max created a PR to fix the missing build flag
